pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -0,0 +1,720 @@
+		}).Debug("finished syncing cluster deployment")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	clusterDeployment, err := c.clusterDeploymentsLister.ClusterDeployments(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Debug("cluster deployment has been deleted")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	clusterDeploymentLog := clustoplog.WithClusterDeployment(c.logger, clusterDeployment)
+
+	// Only attempt to manage clusterdeployments if the version they should run is fully resolvable
+	clusterVersion, err := c.resolveClusterVersion(clusterDeployment)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			c.enqueueAfter(clusterDeployment, 10*time.Second) // recheck for missing version every 10 seconds indefinitely
+			return nil
+		}
+		return err
+	}
+
+	err = validateAWSRegion(clusterDeployment, clusterVersion, clusterDeploymentLog)
+	if err != nil {
+		return c.setInvalidRegionStatus(clusterDeployment, clusterVersion)
+	}
+
+	err = c.syncCluster(clusterDeployment)
+	if err != nil {
+		return err
+	}
+
+	err = c.syncControlPlane(clusterDeployment, clusterVersion)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// syncCluster takes a cluster deployment and ensures that a corresponding cluster exists and that
+// it reflects the spec of the cluster deployment
+func (c *Controller) syncCluster(clusterDeployment *clustop.ClusterDeployment) error {
+	cluster, err := c.clustersLister.Clusters(clusterDeployment.Namespace).Get(clusterDeployment.Name)
+	if err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("cannot retrieve cluster for cluster deployment %s/%s: %v", clusterDeployment.Namespace, clusterDeployment.Name, err)
+	}
+
+	if cluster == nil {
+		// cluster does not exist, it needs to be created
+		cluster, err = buildCluster(clusterDeployment)
+		if err != nil {
+			return fmt.Errorf("cannot build cluster for cluster deployment %s/%s: %v", clusterDeployment.Namespace, clusterDeployment.Name, err)
+		}
+		_, err := c.capiClient.ClusterV1alpha1().Clusters(clusterDeployment.Namespace).Create(cluster)
+		if errors.IsConflict(err) {
+			// cluster was created elsewhere, no need to retry
+			clustoplog.WithClusterDeployment(c.logger, clusterDeployment).Debugf("conflict creating cluster, skipping")
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("error creating cluster for cluster deployment %s/%s: %v", clusterDeployment.Namespace, clusterDeployment.Name, err)
+		}
+		return nil
+	}
+
+	// cluster exists, make sure it reflects the current cluster deployment spec
+	deploymentConfigSpec, err := controller.ClusterDeploymentSpecFromCluster(cluster)
+	if err != nil {
+		return fmt.Errorf("cannot retrieve a cluster deployment spec from existing cluster %s/%s: %v", cluster.Namespace, cluster.Name, err)
+	}
+	if !apiequality.Semantic.DeepEqual(deploymentConfigSpec, clusterDeployment.Spec) {
```

Comment:
deploymentConfigSpec might be worth a rename.

I'd consider a change to the spec worth a log message at info level.